### PR TITLE
[TPU] Implement prefix caching for TPUs

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR implements the prefix caching support for the TPU backend.

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
Nice work!

**Comment by robertgshaw2-neuralmagic** [comment]
In theory, this should enable chunked prefill too right?

File: vllm/attention/backends/pallas.py [hunk 1]
```diff
@@ -186,29 +184,49 @@ def forward(
...
+                output = torch.ops.xla.flash_attention(
+                    query.permute(0, 2, 1, 3),
+                    key.permute(0, 2, 1, 3),
+                    value.permute(0, 2, 1, 3),
+                    True,
+                )
+                output = output.permute(0, 2, 1, 3)
+            else:
+                # Prefill with paged KV cache.
+                # TODO(woosuk): Tune the below knobs.
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/attention/backends/pallas.py:216]: @vanbasten23 Is the fixed kernel available in today's nightly?

[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/attention/backends/pallas.py:217]: @vanbasten23 After the kernel fix, the model generates correct outputs with prefix caching 🎉

[the remark this answers is not in the document]
Comment (vanbasten23) [reply vllm/attention/backends/pallas.py:217]: Awesome. Thanks for confirming!

File: examples/offline_inference_tpu.py [hunk 2]
```diff
@@ -1,28 +1,30 @@
...
+          enable_prefix_caching=True)
+outputs = llm.generate(prompts, sampling_params)
+for output in outputs:
+    prompt = output.prompt
+    generated_text = output.outputs[0].text
+    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
+
 outputs = llm.generate(prompts, sampling_params)
-for output, answer in zip(outputs, answers):
+for output in outputs:
```
Comment (vanbasten23) [inline examples/offline_inference_tpu.py:27]: I wonder if you need a test for the prefix caching.

**Review by vanbasten23 — APPROVED** [review]

**Comment by vanbasten23** [comment]
Btw, which command did you use run `examples/offline_inference_tpu.py`. I used `$ python vllm/examples/offline_inference_tpu.py` but it [fails](https://gist.github.com/vanbasten23/8b0ada503e52318398c6a33dabea624d). Do you need to use a model other than "google/gemma-2b"?

File: vllm/attention/backends/pallas.py [hunk 3]
```diff
@@ -186,29 +185,50 @@ def forward(
...
+                num_kv_pages_per_compute_block = 16
+                num_queries_per_compute_block = 16
+                assert seq_len % num_queries_per_compute_block == 0
+                output = torch.ops.xla.multi_queries_paged_attention(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/attention/backends/pallas.py:221]: @vanbasten23 - does this new kernel have the same SMEM requirements as the original `paged_attention` where the entire block table is stored in SMEM?

E.g. for the decoding run (see below), we split the batch dimension into smaller chunks and run the kernel multiple times

Comment (vanbasten23) [reply vllm/attention/backends/pallas.py:221]: hey @robertgshaw2-neuralmagic , yes this new kernel have the same SMEM requirements. I am aware of the SMEM OOM issue you mentioned and we plan to address it.

**Comment by WoosukKwon** [comment]
@vanbasten23
> Btw, which command did you use run examples/offline_inference_tpu.py. I used $ python vllm/examples/offline_inference_tpu.py but it [fails](https://gist.github.com/vanbasten23/8b0ada503e52318398c6a33dabea624d). Do you need to use a model other than "google/gemma-2b"?

This is weird. Which version & TPU are you using?

**Pull request merged**